File: src/main/java/bad/robot/radiate/Aggregator.java
package bad.robot.radiate;

import bad.robot.radiate.teamcity.Build;
import com.googlecode.totallylazy.Callable2;
import com.googlecode.totallylazy.Sequence;

import static bad.robot.radiate.ActivityAggregator.aggregated;
import static bad.robot.radiate.teamcity.Build.Functions.toActivity;
import static bad.robot.radiate.teamcity.Build.Functions.toStatus;
import static com.googlecode.totallylazy.Sequences.sequence;

public class Aggregator {

    private final Iterable<Build> builds;

    private Aggregator(Iterable<Build> builds) {
        this.builds = builds;
    }

    public static Aggregator aggregate(Iterable<Build> statuses) {
        return new Aggregator(statuses);
    }

    public Activity activity() {
        Sequence<Activity> activities = sequence(builds).map(toActivity());
        return aggregated(activities).getActivity();
    }

    public Status status() {
        Sequence<Status> statuses = sequence(builds).map(toStatus());
        return StatusAggregator.aggregated(statuses).getStatus();
    }

    public Progress progress() {
        Progress seed = sequence(builds).head().getProgress();
        return sequence(builds).tail().fold(seed, (progress, build) -> progress.add(build.getProgress()));
    }


}

File: src/main/scala/bad/robot/radiate/Aggregator.scala
package bad.robot.radiate

import bad.robot.radiate.teamcity.Build

object Aggregator {
  def aggregate(statuses: List[Build]): Aggregator = {
    new Aggregator(statuses)
  }
}

class Aggregator(builds: List[Build]) {

  def activity: Activity = {
    val activities = builds.map(_.activity)
    ActivityAggregator.aggregated(activities).getActivity
  }

  def status: Status = {
    val statuses = builds.map(_.status)
    StatusAggregator.aggregated(statuses).getStatus
  }

  def progress: Progress = {
    builds.tail.foldLeft(builds.head.progress) { (progress, build) => {
      progress.add(build.progress)
    }}
  }
}